util/align_ptr.h

```diff
@@ -0,0 +1,44 @@
+#ifndef _ALIGN_PTR_H_
+#define _ALIGN_PTR_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+/**
+ * ALIGN_PTR_DECL(type, name, num, alignsize):
+ * Declare a pointer called ${name}, which points to an array large enough to
+ * contain ${num} values of ${type} and is aligned to ${alignsize} bytes.  The
+ * pointer must not be used until ALIGN_PTR_INIT(${name}, ${alignsize}) has
+ * been called.  This macro may also create an additional variable called
+ * "${name}_buf".
+ */
+#define ALIGN_PTR_DECL(type, name, num, alignsize)			\
+	uint8_t name##_buf[num * sizeof(type) + (alignsize - 1)];	\
+	type * name
+
+/**
+ * ALIGN_PTR_INIT(name, alignsize):
+ * Initialize the variable called ${name} to point to an array which is
+ * aligned to ${alignsize} bytes.  They must have previously been declared
+ * with ALIGN_PTR_DECL(${name}, ${alignsize}).
```

Comment:
This deserves a comment "Requires that casting a pointer to uintptr_t behaves naturally, i.e. yields a byte offset relative to aligned memory."

The semantics required for uintptr_t are far weaker -- the values don't have to be meaningful at all, beyond returning the same pointer if you cast them back -- but every platform I'm aware of provides this (I think even including CHERI, which breaks all sorts of pointer-related assumptions).